### PR TITLE
Lodash: `flow` (aka `pipe`): reverse order of overloads to fix various issues

### LINKED ISSUE #34379: Lodash: `flow`: incorrect inference for function with optional parameter
As per https://github.com/Microsoft/TypeScript/issues/29913, we currently have this issue when using `flow` (aka `pipe`) **when `strictFunctionTypes` is disabled**:

``` ts
// Expected type: `(a: {} | undefined) => number`
// Actual type: `() => number`
const fn = _.flow((_a?: {}) => 1, x => x);
```

aka

``` ts
declare const pipe: {
    <A>(a: () => A): () => A;
    <A, B>(ab: (a: A) => B): (a: A) => B;
};

// Expected type: `(a: {} | undefined) => number`
// Actual type: `() => number`
const fn = pipe((_a?: {}) => 1);
```

---

If you know how to fix the issue, make a pull request instead.

- [x] I tried using the `@types/lodash` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [ ] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @....

If you do not mention the authors the issue will be ignored.

### LINKED ISSUE #34380: Lodash: `flow`: incorrect inference for generic function
TS v3.4.1

``` ts
declare const pipe: {
    <A, B>(a: () => A, ab: (a: A) => B): () => B;
    <A, B, C>(ab: (a: A) => B, bc: (b: B) => C): (a: A) => C;
};

declare const identity: <T>(t: T) => T;

// Expected and actual type: `<T>(a1: T) => T`
const fn1 = pipe(
    identity,
    identity,
);

// Expected type: `<T>(a1: T) => T`
// Actual type: `(a1: {}) => {}`
const fn2 = pipe(
    identity,
    value => identity(value),
);
```

---

If you know how to fix the issue, make a pull request instead.

- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [ ] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @....

If you do not mention the authors the issue will be ignored.

### PR DESCRIPTION
**Pull request by OliverJAsh**
Fixes https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34379
Fixes https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34380

This is the 2nd attempt. 1st attempt here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33450. (It couldn't be merged due to memory issues when linting, and I'm hoping this is magically fixed now.)

#### To do

- [ ] Update wrappers
- [ ] Update `flowRight`

### CODE REVIEW

File: types/lodash/lodash-tests.ts [hunk 1]
```diff
@@ -3401,10 +3401,22 @@ fp.now(); // $ExpectType number
 // _.flowRight
 {
     const fn1 = (n: number): number => 0;
+    const fn1Optional = (n?: number): number => 0;
     const fn2 = (m: number, n: number): number => 0;
     const fn3 = (a: number): string => "";
     const fn4 = (a: string): boolean => true;
```
Comment (OliverJAsh) [inline types/lodash/lodash-tests.ts]: How can I add a test for the generic function issue (described in this PR's description)? Generics and pipe only work in 3.4.1, but these tests are ran against older versions.

``` ts
declare const identity: <T>(t: T) => T;

// Expected and actual type: `<T>(a1: T) => T`
const fn1 = pipe(
    identity,
    identity,
);

// Expected type: `<T>(a1: T) => T`
// Actual type (prior to this PR): `(a1: {}) => {}`
const fn2 = pipe(
    identity,
    value => identity(value),
);
```

File: types/lodash/common/util.d.ts [hunk 2]
```diff
@@ -237,15 +237,7 @@ declare module "../index" {
...
-        flow<A1, A2, A3, A4, R1, R2, R3, R4, R5>(f1: (a1: A1, a2: A2, a3: A3, a4: A4) => R1, f2: (a: R1) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5): (a1: A1, a2: A2, a3: A3, a4: A4) => R5;
-        flow<A1, A2, A3, A4, R1, R2, R3, R4, R5, R6>(f1: (a1: A1, a2: A2, a3: A3, a4: A4) => R1, f2: (a: R1) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6): (a1: A1, a2: A2, a3: A3, a4: A4) => R6;
-        flow<A1, A2, A3, A4, R1, R2, R3, R4, R5, R6, R7>(f1: (a1: A1, a2: A2, a3: A3, a4: A4) => R1, f2: (a: R1) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6, f7: (a: R6) => R7): (a1: A1, a2: A2, a3: A3, a4: A4) => R7;
-        flow<A1, A2, A3, A4, R1, R2, R3, R4, R5, R6, R7>(f1: (a1: A1, a2: A2, a3: A3, a4: A4) => R1, f2: (a: R1) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6, f7: (a: R6) => R7, ...funcs: Array<Many<(a: any) => any>>): (a1: A1, a2: A2, a3: A3, a4: A4) => any;
-        // any-argument first function
```
Comment (OliverJAsh) [inline types/lodash/common/util.d.ts]: When these tests are bumped to TS version 3.4.1, we can do this instead:

``` ts
        flow<A extends any[], B>(ab: (...args: A) => B): (...args: A) => B;
        flow<A extends any[], B, C>(ab: (...args: A) => B, bc: (b: B) => C): (...args: A) => C;
        flow<A extends any[], B, C, D>(ab: (...args: A) => B, bc: (b: B) => C, cd: (c: C) => D): (...args: A) => D;
        flow<A extends any[], B, C, D, E>(ab: (...args: A) => B, bc: (b: B) => C, cd: (c: C) => D, de: (d: D) => E): (...args: A) => E;
        flow<A extends any[], B, C, D, E, F>(ab: (...args: A) => B, bc: (b: B) => C, cd: (c: C) => D, de: (d: D) => E, ef: (e: E) => F): (...args: A) => F;
        flow<A extends any[], B, C, D, E, F, G>(ab: (...args: A) => B, bc: (b: B) => C, cd: (c: C) => D, de: (d: D) => E, ef: (e: E) => F, fg: (f: F) => G): (...args: A) => G;
        flow<A extends any[], B, C, D, E, F, G, H>(ab: (...args: A) => B, bc: (b: B) => C, cd: (c: C) => D, de: (d: D) => E, ef: (e: E) => F, fg: (f: F) => G, gh: (g: G) => H): (...args: A) => H;
```

https://github.com/Microsoft/TypeScript/issues/29904#issuecomment-467691948

**Comment by OliverJAsh** [comment]
Hi @sandersn, this PR is failing the build due to memory problems when running the tests for Lodash. However, it works fine locally (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33450#issuecomment-468284250). Can you provide any tips for how I can debug this? This is my 2nd attempt at this PR.

**Comment by sandersn** [comment]
@OliverJAsh The parallel runner leaks memory, I'm pretty sure. I should probably switch it to spawning subprocesses the way that the overnight run does. Since you can specify that the local run should have more memory, I think that should be as good as passing the CI checks.

Before checking in, though, I remember that lodash is already near the limit of causing memory problems in real-world usage. Can you check a couple of performance things before we check this in?

1. Try this change with a real, preferably large, project to make sure that it doesn't run out of memory there.
2. Compare the resource usage with-vs-without the change by passing `--extendedDiagnostics` to tsc.

**Comment by jgornick** [comment]
@OliverJAsh Would this fix utilize TypeScript 3.4 [higher order type inference from generic functions](https://devblogs.microsoft.com/typescript/announcing-typescript-3-4/#higher-order-type-inference-from-generic-functions), allowing composition of generic React higher-order components?

I've run into a case when using `flow` to compose [`react-redux`](https://github.com/reduxjs/react-redux).[`connect`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/1eb3865ae7e39366f4e80c5f180450055465430b/types/react-redux/v6/index.d.ts#L258) and [`react-i18next`](https://github.com/i18next/react-i18next).[`translate`](https://github.com/i18next/react-i18next/blob/11e23a4c482a3449ca1e664eedd745975957e3d6/src/index.d.ts#L70), that only the return type from translate are being inferred, but not that from `connect`.

Am I incorrectly using `flow` here?

Example:

```
export const MyComponent = flow(
  connect(mapStateToProps, mapDispatchToProps),
  translate('common')
)(MyReactComponent)
```

Thank you!

**Comment by OliverJAsh** [comment]
@jgornick No this wouldn't fix that. See https://github.com/Microsoft/TypeScript/issues/30650

**Comment by OliverJAsh** [comment]
@jgornick Actually I was wrong, that issues only concerns generic components, which I'm guessing isn't your case.

**Comment by OliverJAsh** [comment]
@jgornick I would file a separate issue with steps to reproduce. Ping me and I'm happy to glance over it.

**Comment by OliverJAsh** [comment]
> 1. Try this change with a real, preferably large, project to make sure that it doesn't run out of memory there.

I'll give this a shot.

In any case this PR is on hold, because I realised that this change fixes the examples I provided, but breaks something else:

``` ts
declare const pipe: {
    <A, B, C>(ab: (a: A) => B, bc: (b: B) => C): (a: A) => C;
    <A, B>(a: () => A, ab: (a: A) => B): () => B;
};

{
    // Expected type: `() => string`
    // Actual type: `(a: {}) => string`
    const fn1 = pipe(
        () => 'foo',
        x => x,
    );
}
```

I'm not sure how to create a happy medium here—something that preserves the behaviour above whilst also fixing the issues I mentioned (in this PR's description and in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33450). I asked the TS team here: https://github.com/Microsoft/TypeScript/issues/29904#issuecomment-478720896. The only solution I currently know of is generic rest parameters, but they require a newer TS version than what the Lodash typings currently support. If anyone has any ideas, I'd love to hear them.

**Comment by OliverJAsh** [comment]
I decided to go ahead and remove the 0-argument first function overloads, since it's the only way I know to fix https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34379 and https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34380.

This is a breaking change.

Usage before:

``` ts
// () => string
const fn1 = pipe(
    () => 'foo',
    x => x,
);

fn1()
```

Usage after:

``` ts
// (a1: {}) => string
const fn1 = pipe(
    () => 'foo',
    x => x,
);

// must pass empty object in here
fn1({})
```

Please could someone review this PR now? If we're happy with the changes, I will update wrappers and `flowRight`. /cc @bczengel @chrootsu @stepancar @aj-r @Ailrun @e-cloud @thorn0 @jtmthf @DomiR

**Comment by OliverJAsh** [comment]
This PR is failing the build due to memory problems when running the tests for Lodash. I need to debug this (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34374#issuecomment-478660035), but before I try to do that, please can I get a review on the code changes? Once the theory is approved, I'll update the wrappers + `flowRight` and _then_ I'll debug the build failure.

**Comment by aj-r** [comment]
Sorry for the late response - I don't think that removing the 0-argument overload is the right thing to do, though. I don't think it's worth it to break the 0-argument overload for everybody, in order to fix an edge case when `strictFunctionTypes` is `false`. I feel like the solution here is to change `strictFunctionTypes` to true rather than change the lodash types.

**Comment by OliverJAsh** [comment]
Thanks for the response @aj-r. This change isn't just concerning bugs related to `strictFunctionTypes`. In addition to #34379, see #34380.

**Pull request closed without merging**